app/scripts/modules/core/src/overrideRegistry/Overridable.tsx

```diff
@@ -0,0 +1,144 @@
+ */
+export function overridableComponent <P extends IOverridableProps, T extends React.ComponentClass<P>> (DefaultComponent: T, key: string): T {
+  class OverridableComponent extends React.Component<P, { Component: T }> {
+    private account$ = new Subject<string>();
+    private destroy$ = new Subject();
+
+    constructor(props: P) {
+      super(props);
+
+      const { accountService } = ReactInjector;
+
+      this.account$
+        .switchMap(accountId => {
+          if (accountId) {
+            return Observable.fromPromise(accountService.getAccountDetails(accountId));
+          }
+
+          return Observable.of(null);
+        })
+        .map((accountDetails: IAccountDetails) => this.getComponent(accountDetails))
+        .takeUntil(this.destroy$)
+        .subscribe((Component) => this.setState({ Component }));
+    }
+
+    public componentDidMount() {
+      this.account$.next(this.props.accountId);
+    }
+
+    public componentWillUnmount() {
+      this.destroy$.next();
+    }
+
+    public componentWillReceiveProps(nextProps: P) {
+      const { accountId } = nextProps;
+      if (this.props.accountId !== accountId) {
+        this.account$.next(accountId);
+      }
+    }
+
+    private getComponentFromCloudProvider(accountDetails: IAccountDetails): T {
+      const { cloudProvider, providerVersion } = accountDetails;
+      const { cloudProviderRegistry } = ReactInjector;
+      if (!cloudProvider) {
+        return null;
+      }
+
+      const CloudProviderComponentOverride = cloudProviderRegistry.getValue(cloudProvider, key, providerVersion);
+      if (CloudProviderComponentOverride) {
+        return CloudProviderComponentOverride as T;
+      }
+
+      const cloudProviderTemplateOverride = cloudProviderRegistry.getValue(cloudProvider, key + 'TemplateUrl', providerVersion);
+      if (cloudProviderTemplateOverride) {
+        const cloudProviderController = cloudProviderRegistry.getValue(cloudProvider, key + 'Controller', providerVersion);
+        const controllerAs = cloudProviderController && cloudProviderController.includes(' as ') ? undefined : 'ctrl';
+        const Component = (props: any) =>
+          <AngularJSAdapter {...props} templateUrl={cloudProviderTemplateOverride} controller={cloudProviderController} controllerAs={controllerAs}/>;
+
+        return Component as any as T;
+      }
+
+      return null;
+    }
+
+    private getComponentFromOverrideRegistry(): T {
+      const { overrideRegistry } = ReactInjector;
+
+      const ComponentOverride = overrideRegistry.getComponent(key);
+      if (ComponentOverride) {
+        return ComponentOverride as T;
+      }
+
+      const templateOverride: string = overrideRegistry.getTemplate(key, null);
+      if (templateOverride) {
+        const controllerOverride: string = overrideRegistry.getController(key, null);
+        const controllerAs = controllerOverride && controllerOverride.includes(' as ') ? undefined : 'ctrl';
```

Comment:
If the controller string doesn't include an `as` (i.e., `FooController as $ctrl`), then use `ctrl`.  This is a convention, I suppose.